snow/engine/avalanche/bootstrap/bootstrapper.go

```diff
@@ -346,24 +355,41 @@ func (b *Bootstrapper) ForceAccepted(acceptedContainerIDs []ids.ID) error {
 	return b.process(toProcess...)
 }
 
-// Finish bootstrapping
-func (b *Bootstrapper) finish() error {
+// checkFinish repeatedly executes pending transactions and requests new frontier blocks until there isn't any new ones
+// after which it finishes the bootstrap process
+func (b *Bootstrapper) checkFinish() error {
 	// If there are outstanding requests for vertices or we still need to fetch vertices, we can't finish
 	if b.Ctx.IsBootstrapped() || b.OutstandingRequests.Len() > 0 || b.needToFetch.Len() > 0 {
 		return nil
 	}
 
-	b.Ctx.Log.Info("bootstrapping fetched %d vertices. executing transaction state transitions...",
+	b.Ctx.Log.Info("bootstrapping fetched %d blocks. executing state transitions...",
 		b.NumFetched)
-	if err := b.executeAll(b.TxBlocked, b.Ctx.DecisionDispatcher); err != nil {
+
+	executedBlocks, err := b.executeAll(b.TxBlocked, b.Ctx.DecisionDispatcher)
+	if err != nil {
 		return err
 	}
 
 	b.Ctx.Log.Info("executing vertex state transitions...")
-	if err := b.executeAll(b.VtxBlocked, b.Ctx.ConsensusDispatcher); err != nil {
+	executedVertices, err := b.executeAll(b.VtxBlocked, b.Ctx.ConsensusDispatcher)
+	if err != nil {
 		return err
 	}
 
+	if executedBlocks+executedVertices > b.executedStateTransitions && executedBlocks+executedVertices > 0 && b.RetryBootstrap {
```

Comment:
Do we want to have some threshold here `> 0` so that we don't get stuck looping when there is high load?

Reply:
If we wanted to prevent needless traversals, we should probably do something like `executedVertices < .9 * b.executedStateTransitions`

Reply:
Something like this ?

```go
        // exec - 10k ; prev - 0     ; runs again because prev is math.MaxInt32
	// exec - 20  ; prev - 10k/2 ; runs again
	// exec - 2   ; prev - 20/2  ; runs again
	// exec - 1   ; prev - 2/1   ; doesnt run again
	//
	previousStateTransitionsThreshold := b.executedStateTransitions
	if previousStateTransitionsThreshold == 0 {
		previousStateTransitionsThreshold = math.MaxInt32
	} else {
		previousStateTransitionsThreshold /= 2
	}

	if executedVertices < previousStateTransitionsThreshold && b.RetryBootstrap {
		b.executedStateTransitions = executedVertices
		b.Ctx.Log.Info("bootstrapping is checking for more vertices before finishing the bootstrap process...")
		return b.RestartBootstrap(true)
	}
```